### PR TITLE
fix malformed body and include additional optional data

### LINKED ISSUE #1068: MultiPart Request Error
The MultiPart body is malformed, the string value contains an additional **\r\n** and the server doesn't accept it when the values are validated.
Also, when uploading a file, "filename" is required by Django, but there is no way to provide additional keys and values to the message part.

### PR DESCRIPTION
**Pull request by p30arena**
Fixes #1068

### CODE REVIEW

**Comment by p30arena** [comment]
@microsoft-github-policy-service agree company="Donusoft"

File: packages/abstractions/src/serialization/serializationWriter.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ export interface SerializationWriter {
    * @param key the key to write the value with.
    * @param value the value to write to the stream.
    */
-  writeStringValue(key?: string | undefined, value?: string | undefined): void;
```
Comment (baywet) [inline packages/abstractions/src/serialization/serializationWriter.ts:22]: please revert this change, we want the serialization writer interfaces to be consistent across languages

File: packages/serialization/multipart/src/multipartSerializationWriter.ts [hunk 2]
```diff
@@ -42,7 +42,9 @@ export class MultipartSerializationWriter implements SerializationWriter {
       }
       this.writeRawStringValue(value);
     }
-    this.writeRawStringValue("\r\n");
```
Comment (baywet) [inline packages/serialization/multipart/src/multipartSerializationWriter.ts:45]: do you think removing this line instead, and adding `\r\n` to the calls to writeStringValue that require it would be a viable alternative here?

File: packages/abstractions/src/multipartBody.ts [hunk 3]
```diff
@@ -128,13 +131,18 @@ export function serializeMultipartBody(
     writer.writeStringValue(undefined, "--" + boundary);
     const part = parts[partName];
     writer.writeStringValue("Content-Type", part.contentType);
+    let dispositionValue = 'form-data; name="' + part.originalName + '"';
```
Comment (baywet) [inline packages/abstractions/src/multipartBody.ts:134]: can you add additional tests here https://github.com/microsoft/kiota-typescript/blob/main/packages/serialization/multipart/test/common/multipartSerializationWriter.ts for that additional behaviour please?

**Review by baywet — CHANGES_REQUESTED** [review]
Thanks for the contribution! A couple of remarks to ensure this gets merged swiftly.

File: packages/abstractions/src/serialization/serializationWriter.ts (hunk 1 above)
Comment (p30arena) [reply packages/abstractions/src/serialization/serializationWriter.ts:22]: I agree

File: packages/serialization/multipart/src/multipartSerializationWriter.ts (hunk 2 above)
Comment (p30arena) [reply packages/serialization/multipart/src/multipartSerializationWriter.ts:45]: can't we make writeRawStringValue public?
if so I could replace this line with:
writeRawStringValue

Comment (baywet) [reply packages/serialization/multipart/src/multipartSerializationWriter.ts:45]: ideally we wouldn't. The idea is that the Serializer implementations only expose the methods defined in the interfaces.

File: packages/abstractions/src/multipartBody.ts (hunk 3 above)
Comment (p30arena) [reply packages/abstractions/src/multipartBody.ts:134]: addOrReplacePart changes are not consistent across languages, isn't it a problem?

Comment (p30arena) [reply packages/abstractions/src/multipartBody.ts:134]: https://www.ietf.org/rfc/rfc2388.txt

it seems that filename is not mandatory.

> The original local file name may be supplied as well, either as a
>    "filename" parameter either of the "content-disposition: form-data"
>    header or, in the case of multiple files, in a "content-disposition:
>    file" header of the subpart. The sending application MAY supply a
>    file name; if the file name of the sender's operating system is not
>    in US-ASCII, the file name might be approximated, or encoded using
>    the method of RFC 2231.

Comment (baywet) [reply packages/abstractions/src/multipartBody.ts]: > addOrReplacePart changes are not consistent across languages, isn't it a problem?

Can you provide more context for this please?

**Comment by p30arena** [comment]
@baywet so what's the status?

**Comment by baywet** [comment]
@p30arena sorry I've been out for medical reasons most of this week.
I'm waiting for two things:
- resolution of this conversation https://github.com/microsoft/kiota-typescript/pull/1069#discussion_r1487881255
- added unit tests to prevent regressions

**Comment by shemogumbe** [comment]
Closing in favor of https://github.com/microsoft/kiota-typescript/pull/1423
